// Copyright (c) 2014, Cloudera, inc.
// Confidential Cloudera Information: Covered by NDA.
//
// Protobuf used for introspection of RPC services (eg listing in-flight RPCs,
// reflection, etc)

package kudu.rpc;

option java_package = "org.kududb";

import "kudu/rpc/rpc_header.proto";

message RpcCallInProgressPB {
  required RequestHeader header = 1;
  optional string trace_buffer = 2;
  optional uint64 micros_elapsed = 3;
}

message RpcConnectionPB {
  enum StateType {
    UNKNOWN = 999;
    NEGOTIATING = 0;  // Connection is still being negotiated.
    OPEN = 1;         // Connection is active.
  };

  required string remote_ip = 1;
  required StateType state = 2;
  // TODO: swap out for separate fields
  optional string remote_user_credentials = 3;
  repeated RpcCallInProgressPB calls_in_flight = 4;
}

message DumpRunningRpcsRequestPB {
  optional bool include_traces = 1 [ default = false ];
}

message DumpRunningRpcsResponsePB {
  repeated RpcConnectionPB inbound_connections = 1;
  repeated RpcConnectionPB outbound_connections = 2;
}
